src/inmanta/agent/agent.py

```diff
@@ -1672,39 +1672,54 @@ async def get_code(
+                if self._last_loaded_version[resource_install_spec.resource_type] == resource_install_spec.model_version:
+                    LOGGER.debug(
+                        "Handler code already installed for %s version=%d",
+                        resource_install_spec.resource_type,
+                        resource_install_spec.model_version,
+                    )
                     continue
-            try:
-                # Install required python packages and the list of ``ModuleSource`` with the provided pip config
-                await self._install(
-                    resource_install_spec.sources,
-                    resource_install_spec.requirements,
-                    pip_config=resource_install_spec.pip_config,
-                )
-                LOGGER.debug(
-                    "Installed handler %s version=%d", resource_install_spec.resource_type, resource_install_spec.model_version
-                )
-                # Update the ``_last_loaded`` cache to indicate that the given resource type's code
-                # was loaded successfully at the specified version.
-                self._last_loaded[(resource_install_spec.resource_type, resource_install_spec.model_version)] = (
-                    resource_install_spec
-                )
-            except Exception:
-                LOGGER.exception(
-                    "Failed to install handler %s version=%d",
-                    resource_install_spec.resource_type,
-                    resource_install_spec.model_version,
-                )
-                failed_to_load.add(resource_install_spec.resource_type)
+
+                self._last_loaded_version[resource_install_spec.resource_type] = -1
+
+                try:
+                    # Install required python packages and the list of ``ModuleSource`` with the provided pip config
+                    LOGGER.debug(
+                        "Installing handler %s version=%d",
+                        resource_install_spec.resource_type,
+                        resource_install_spec.model_version,
+                    )
+                    await self._install(
+                        resource_install_spec.sources,
+                        resource_install_spec.requirements,
+                        pip_config=resource_install_spec.pip_config,
+                    )
+                    LOGGER.debug(
+                        "Installed handler %s version=%d",
+                        resource_install_spec.resource_type,
+                        resource_install_spec.model_version,
+                    )
+                    # Update the ``_previously_loaded`` cache to indicate that the given resource type's code
+                    # was loaded successfully at the specified version.
+                    self._previously_loaded[(resource_install_spec.resource_type, resource_install_spec.model_version)] = (
+                        resource_install_spec
+                    )
```

Comment:
I would think that this one is the responsibility of `get_code` rather than `ensure_code`? i.e. `get_code` fetches the code and manages its cache, `ensure_code` loads code and keeps track of what is currently loaded.